### DISCUSSION TITLE
Alpine js validate and update x-model before it changes and should take effect

### DISCUSSION THREAD

**Discussion by Shashank-Bhatt1** [alpinejs/alpine#3596] [1. Help] [answered]
Hi, I need one help where I need to validate user input whether he entered a number which is not larger than max value but could not get it to work properly. 

It seems `min` and `max` values entered in html are not handled via min and max attribute so need to reset manually via Javascript.  It just obeys arrow up and down function but not in case when someone enters number there manually.

I am working on one color picker where i need to make sure that someone doesn't enter value larger than 255 and not lower than 0 in order to get red value in rgb color mode, as that value will again be used in calculating hsl values to update `hue slider` which is used to update hue part.

As we can see in this fiddle i have tried to call `updateRed` method and it kind of works but when i have multiple inputs for green and blue, I need this kind of validation before further processing of their respective `x-model's`

https://jsfiddle.net/hxvL0dtw/ 

If I later try to update value in method then red will not be updated as we know that it is not directly bound to red and it's just parameter of the function.

i.e. I am not able to write `updateRed(red)` and inside method

```
if(value != '') {
      if(parseInt(value) < 0) {
        value = 0;
      } 
      if(parseInt(value) > 255) {
        value = 255;
      }
    }
```
because this only updates value and not red itself. but i need this validation in just one function for all red, green and blue.... I also tried `x-modelable` but not worked. 

As we can see in fiddle, it works for red but not for green as i need only one validation function in x-data for all of them.

Any suggestion?

**Comment by ekwoka** [comment]
use a getter/setter pair around an internal value.

```js
{
  _value: 0;
  get value() {
    return this._value
  },
  set value(val) {
    this._value = Math.min(255, Math.max(0, val))
  }
}
```

Then x-model value

**Reply by Shashank-Bhatt1** [reply]
Hi Thanks for looking into this.
So should I take getter and setter for every property like red, green and blue? Do you mean like this? I am not getting what `_value` represents here.

```
red: 0,
get red() {
    return this.red
 },
set red(val) {
    this.red= Math.min(255, Math.max(0, val))
 }
```
Also after taking like above, do I need to create separate getter and setter for for every property type like `get green()` and `get blue()`?

**Reply by ekwoka** [reply, marked as answer]
`_value` is just a placeholder for the internal state.

You can't have the getter setter and the value have the same names. prefixing with `_` is a common way to indicate private/internal values on objects.

**Reply by Shashank-Bhatt1** [reply]
Ah okay now I got it what it means as earlier I didn't know about the actual use case of getter and setter till yet.  So also now got aware after having this. Thanks for the explanation. 

Also, just wanted to mention that although above trick works, I just found one more  thing on alpine's dev website called `programmatic access of x-model` which I believe can also be used somehow.
https://alpinejs.dev/directives/model#programmatic%20access

**Reply by ekwoka** [reply]
The above can work if you want to have the new value go through the pipeline provided by x-model, like with the modifiers `.number` etc.

It doesn't really impact what you'd do for this use case.
